### ISSUE TITLE
Indentation of window functions

### ISSUE THREAD

**Issue by alanmcruickshank** [sqlfluff/sqlfluff#4033]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

As mentioned in #1839 by @alexscott-ff, the following query is currently acceptable by the indentation engine:

```sql
SELECT
    field_one,
    -- Shouldn't line continuations be indented?
    min(event_tracking_activity_id)
    OVER (PARTITION BY some_id_used_to_partition) = event_tracking_activity_id
    AS is_first_val
FROM
    my_table
```

This feels strange because the `OVER` clause isn't indented.

### Expected Behaviour

I would expect a better indentation to be:

```sql
SELECT
    min(event_tracking_activity_id)
        OVER (PARTITION BY some_id_used_to_partition) = event_tracking_activity_id
        AS is_first_val
FROM
    my_table
```

which I think is also congruent with an indentation for normal functions as follows:

```sql
SELECT
    min(event_tracking_activity_id)
        AS is_first_val
FROM
    my_table
```

The `Indent` and `Dedent` tokens should be reviewed to achieve the desired outcome.

### Observed Behaviour

see above

### How to reproduce

lint using #3942

### Dialect

ansi

### Version

main

### Configuration

ansi, L003 only

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by alexscott-ff** [comment]
Hi @alanmcruickshank thanks for creating this; I just happened upon it again after trying `1.4.5` and seeing the issues elaborated on above are still present. Do you have any idea if this will be worked on soon? We've had to ignore `L003` for now, which is kinda a biggy so not ideal.

**Comment by alanmcruickshank** [comment]
I've been digging into this issue to revise indentation here and it has raised a set of other issues. I'm posting this to document the investigations I've done. When functions are short it's fairly easy, but as they get longer I think it becomes harder to see the obvious route. I'll use a example to illustrate:

```sql
-- Lets call this `raw_query`
SELECT
    my_function(col1 + col2, arg2, arg3) over (partition by col3, col4 order by col5 rows between unbounded preceding and current row) as my_relatively_long_alias,
    my_other_function(col6, col7 + col8, arg4) as my_other_relatively_long_alias,
    my_expression_function(col6, col7 + col8, arg4) = col9 + col10 as another_relatively_long_alias
FROM my_table
```

### Status Quo

Currently (as at writing, the `main` branch is at f13a81f4b0e3df78ff917b3826eec740225b6bea), the suggested fix (with the default line length of 80 and using `LT05`) to this is:

```sql
-- Lets call this `fix_1`
SELECT
    my_function(
        col1 + col2, arg2, arg3
    ) over (
        partition by col3, col4
        order by col5 rows between unbounded preceding and current row
    ) as my_relatively_long_alias,
    my_other_function(
        col6, col7 + col8, arg4
    ) as my_other_relatively_long_alias,
    my_expression_function(col6, col7 + col8, arg4)
    = col9 + col10 as another_relatively_long_alias
FROM my_table
```

A few things to note:
1. No newlines before alias expressions (i.e. `AS` clauses). Instead, the alias comes on the same line as the last line of the expression.
2. We're splitting on bracketed expressions rather than on the elements of the window expression.

The potentially unexpected behaviour comes when the initial query contains line breaks in the places described in the original query. If I apply the formatting from the first query in this issue:

```sql
-- This would be the original query with line breaks before `OVER` and `AS`. Lets call this `with_newlines`.
SELECT
    my_function(col1 + col2, arg2, arg3)
    over (partition by col3, col4 order by col5 rows between unbounded preceding and current row)
    as my_relatively_long_alias,
    my_other_function(col6, col7 + col8, arg4)
    as my_other_relatively_long_alias,
    my_expression_function(col6, col7 + col8, arg4) = col9 + col10
    as another_relatively_long_alias
FROM my_table;

-- Which is fixed to:
SELECT
    my_function(col1 + col2, arg2, arg3)
    over (
        partition by col3, col4
        order by col5 rows between unbounded preceding and current row
    )
    as my_relatively_long_alias,
    my_other_function(col6, col7 + col8, arg4)
    as my_other_relatively_long_alias,
    my_expression_function(col6, col7 + col8, arg4) = col9 + col10
    as another_relatively_long_alias
FROM my_table;
```

In this example:
1. `LT05` doesn't complain about the later lines because they're already short enough.
2. `LT02` doesn't try and reindent the later lines because it doesn't think there should be an indent before `OVER` or `AS`.

### What if we indent before `AS`?

One fairly easy change would be to add an indent before `AS` (i.e. in `AliasExpressionSegment`). Note, doing this exposed a bug in the reindent logic (see #4562) and also in the handling of trailing commas (see #4563), so from here onwards, testing will assume that both are fixed. If we do this, then the `raw_query` example which I describe first is fixed in _exactly the same way_, however the second example (the one I called `with_newlines`) is then fixed to this:

```sql
SELECT
    my_function(col1 + col2, arg2, arg3)
    over (
        partition by col3, col4
        order by col5 rows between unbounded preceding and current row
    )
        as my_relatively_long_alias,
    my_other_function(col6, col7 + col8, arg4)
        as my_other_relatively_long_alias,
    my_expression_function(col6, col7 + col8, arg4) = col9 + col10
        as another_relatively_long_alias
FROM my_table;
```

Things to note here:
1. The placement of commas appears a bit strange (suggesting a potential exception required for dedents before commas - which is fixable similar to semicolons).
2. The placement of the `as` clause in the second and third functions seems appropriate, but looks a little strange in the first one (because of the lack of indent before `over`).

### What if we _also_ indent before `OVER`?

Let's _also_ add an indent before `OVER` (i.e. at the ends of `OverClauseSegment`). That leads to the following fix for the original query (the one called `raw_query`):

```sql
SELECT
    my_function(
        col1 + col2, arg2, arg3
    )
        over (
            partition by col3, col4
            order by col5 rows between unbounded preceding and current row
        )
        as my_relatively_long_alias
    ,
    my_other_function(
        col6, col7 + col8, arg4
    ) as my_other_relatively_long_alias,
    my_expression_function(col6, col7 + col8, arg4)
    = col9 + col10 as another_relatively_long_alias
FROM my_table
```

...and this fix for `with_newlines`:

```sql
SELECT
    my_function(
        col1 + col2, arg2, arg3
    )
        over (
            partition by col3, col4
            order by col5 rows between unbounded preceding and current row
        )
        as my_relatively_long_alias,
    my_other_function(
        col6, col7 + col8, arg4
    ) as my_other_relatively_long_alias,
    my_expression_function(col6, col7 + col8, arg4)
    = col9 + col10 as another_relatively_long_alias
FROM my_table
```

Note that they're **the same!** 🎉. Interesting, the splitting of `my_function(...)` is unnecessary, but this is a limitation of the long line fixing algorithm which could be fairly easily improved.

### Plan forward.

- First we need to merge #4562 & #4563.
- I suggest that I also address the long line fixing algorithm to make it a little smarter first.
- Then we can add these indents and dedents in the right places.
